documentation/high-avail/ha-api-hld.md

```diff
@@ -0,0 +1,617 @@
+    SAI_DASH_HA_FLOW_SYNC_STATE_FLOW_PENDING_RESIMULATION
+} sai_dash_ha_flow_sync_state_t;
+```
+
+The flow sync state is a small state machine that represents if the flow is synched or not, so we can make the packet processing decision accordingly to achieve HA.
+
+```mermaid
+stateDiagram-v2
+    M: FLOW_MISS
+    C: FLOW_CREATED
+    S: FLOW_SYNCED
+    D: FLOW_PENDING_DELETE
+    R: FLOW_PENDING_RESIMULATION
+
+    M --> C: Flow created
+    M --> S: Flow sync request<br>received on<br>standby node
+
+    C --> S: Flow sync ack received
+    S --> R: Flow resimulation requested
+    R --> C: Flow updated
+
+    C --> D: Flow deleted
+    S --> D: Flow deleted
+    R --> D: Flow deleted
+
+    D --> M: Flow sync ack received
+```
+
+For more information, please refer to DASH flow API documentation.
+
+### 4.5. ENI
+
+To provide the ENI-level HA control, each ENI will have the following SAI attributes:
+
+| Attribute name | Type | Description |
+| -------------- | ---- | ----------- |
+| SAI_ENI_ATTR_HA_SCOPE_ID | `sai_object_id_t` | The HA scope ID of the ENI. |
+
+### 4.6. Event notifications
+
+To receive the HA related updates from the DASH implementation, the following SAI notification attributes are added on the switch object:
+
+| Attribute name | Type | Description |
+| -------------- | ---- | ----------- |
+| SAI_SWITCH_ATTR_HA_SET_EVENT_NOTIFY | `sai_ha_set_event_notification_fn` | The callback function for receiving events on the HA set. |
+| SAI_SWITCH_ATTR_HA_SCOPE_EVENT_NOTIFY | `sai_ha_scope_event_notification_fn` | The callback function for receiving events on the HA scope. |
+
+#### 4.6.1. HA set event notifications
+
+Whenever a HA set state is changed, it will be reported back via HA set event notification, such as data plane channel goes down. The detailed definition is shown as below:
+
+```c
+/**
+ * @brief HA set event type
+ */
+typedef enum _sai_ha_set_event_t
+{
+    /** Any HA set state is changed, such as data plane channel goes down. */
+    SAI_HA_SET_STATE_CHANGED,
+
+} sai_ha_set_event_t;
+
+/**
+ * @brief Notification data format received from SAI HA set callback
+ *
+ * @count attr[attr_count]
+ */
+typedef struct _sai_ha_set_event_data_t
+{
+    /** Event type */
+    sai_ha_set_event_t event_type;
+
+    /** HA set id */
+    sai_object_id_t ha_set_id;
+
+    /** Attributes count */
+    uint32_t attr_count;
+
+    /**
+     * @brief Attributes
+     *
+     * @objects SAI_OBJECT_TYPE_HA_SET
+     */
+    sai_attribute_t *attr;
+
+} sai_ha_set_event_data_t;
+
+/**
+ * @brief HA set event notification
+ *
+ * Passed as a parameter into sai_initialize_switch()
+ *
+ * @count data[count]
+ *
+ * @param[in] count Number of notifications
+ * @param[in] data Array of HA set events
+ */
+typedef void (*sai_ha_set_event_notification_fn)(
+        _In_ uint32_t count,
+        _In_ const sai_ha_set_event_data_t *ha_set_event_data);
+```
+
+#### 4.6.2. HA scope event notifications
```

Comment:
Seems redundant, HA scope does not define liveliness API for this event, it's in the HA_SET.

Reply:
I see, the HA scope state change event is used for reporting the HA state changes. For example, whenever we call SAI API to update the HA role to "Active". It might take some time for the implementation to process the request and really switch to active. Then whenever it is completed, we can send a HA scope event with the HA role set to "Active" in its event data.

This also helps the case where DPU decides to drive the HA state machine without using NPU ([HLD here](https://github.com/r12f/SONiC/blob/user/r12f/ha2/doc/smart-switch/high-availability/smart-switch-ha-dpu-scope-dpu-driven-setup.md)). Whenever the HA role is changed by DPU, it will be reported using this event notification as well.

Reply:
Setting HA scope role is synchronous. the set_attribute call should not return until it is completed.

Reply:
yea, it could be synchronous for the first case. But for the second case, this notification will be required, otherwise we won't be able to know the state being changed.